Source/Plugins/Core/com.equella.core/js/tsrc/search/SearchPageHelper.ts

```diff
@@ -0,0 +1,376 @@
+        start: primaryDate,
+        end: primaryDate,
+      }),
+    ]
+  )(rangeType.toLowerCase() as RangeType);
+};
+
+const getDisplayModeFromLegacyParams = (
+  legacyDisplayMode: string | undefined
+): DisplayMode =>
+  pipe(
+    legacyDisplayMode,
+    match(
+      // When type is 'standard' or undefined, default to 'list'.
+      [Union(Literal("standard"), Undefined), (): DisplayMode => "list"],
+      [Literal("gallery"), (): DisplayMode => "gallery-image"],
+      [Literal("video"), (): DisplayMode => "gallery-video"],
+      [
+        Unknown,
+        () => {
+          throw new Error("Unknown Legacy display mode");
+        },
+      ]
+    )
+  );
+
+const getCollectionFromLegacyParams = async (
+  collectionUuid: string | undefined
+): Promise<Collection[] | undefined> => {
+  if (!collectionUuid) return defaultSearchOptions.collections;
+  const collectionDetails:
+    | Collection[]
+    | undefined = await findCollectionsByUuid([collectionUuid]);
+
+  return typeof collectionDetails !== "undefined" &&
+    collectionDetails.length > 0
+    ? collectionDetails
+    : defaultSearchOptions.collections;
+};
+
+const getOwnerFromLegacyParams = async (ownerId: string | undefined) =>
+  ownerId ? await findUserById(ownerId) : defaultSearchOptions.owner;
+
+/**
+ * A function that takes query string params from a shared searching.do URL and converts all applicable params to SearchPageOptions.
+ * @param params URLSearchParams from a shared `searching.do` URL
+ * @return SearchPageOptions object.
+ */
+export const legacyQueryStringToSearchPageOptions = async (
+  params: URLSearchParams
+): Promise<SearchPageOptions> => {
+  const getQueryParam = (paramName: LegacyParams) => {
+    return params.get(paramName) ?? undefined;
+  };
+  const query = getQueryParam("q") ?? defaultSearchOptions.query;
+  const collections = await getCollectionFromLegacyParams(
+    getQueryParam("in")?.substring(1)
+  );
+  const owner = await getOwnerFromLegacyParams(getQueryParam("owner"));
+  const sortOrder = pipe(
+    getQueryParam("sort")?.toUpperCase(),
+    O.fromPredicate(OEQ.SearchSettings.SortOrderRunTypes.guard),
+    O.fold(
+      () => defaultSearchOptions.sortOrder,
+      (s) => s
+    )
+  );
+
+  const datePrimary = getQueryParam("dp");
+  const dateSecondary = getQueryParam("ds");
+  const lastModifiedDateRange =
+    getLastModifiedDateRangeFromLegacyParams(
+      getQueryParam("dr") as RangeType,
+      datePrimary ? new Date(parseInt(datePrimary)) : undefined,
+      dateSecondary ? new Date(parseInt(dateSecondary)) : undefined
+    ) ?? defaultSearchOptions.lastModifiedDateRange;
+
+  const mimeTypeFilters = await getMimeTypeFiltersById(params.getAll("mt"));
+  const mimeTypes = mimeTypeFilters?.flatMap(({ mimeTypes }) => mimeTypes);
+  const getExternalMIMETypes = () => {
+    const integrationMIMETypes = params.getAll("_int.mimeTypes");
+    return integrationMIMETypes.length > 0 ? integrationMIMETypes : undefined;
+  };
```

Comment:
Rather than setting up a function, how's about:

```suggestion
  const externalMimeTypes = pipe(
    params.getAll("_int.mimeTypes"),
    O.fromPredicate(A.isNonEmpty),
    O.toUndefined
  );
```

Then below you can just use it directly for the object in the `return` statement.

Reply:
Good idea!  I feel the name could be better if it was `O.orUndefined` . :smile:

Reply:
Yeah, I agree. It's a bit awkward that name. Probably should have gone with convention more and use `O.getOrElseUndefined` and `O.getOrElseNull`. As in the end, it's only needed as you can't returned `undefined` (and I guess `null` but didn't try) from `O.getOrElse`.